### DISCUSSION TITLE
Why renaming is not allowed for non-empty directories using cloud-based filesystems?

### DISCUSSION THREAD

**Discussion by mezuky04** [drakkan/sftpgo#1244] [General]
Hello. Recently I noticed that some folders cannot be renamed. In the API documentation is stated that `The rename is allowed for empty directory or for non empty local directories, with no virtual folders inside`. The folders for which the rename fails are indeed using the s3 filesystem. Is there a technical limitation that prevents the renaming of directories using cloud-based filesystems? Thanks

**Comment by drakkan** [comment]
Hello,

in the development version I have added this configuration parameter

>  - `rename_mode`, integer. By default (`0`), renaming of non-empty directories is not allowed for cloud storage providers (S3, GCS, Azure Blob). Set to `1` to enable recursive renames for these providers, they may be slow, there is no atomic rename API like for local filesystem, so SFTPGo will recursively list the directory contents and do a rename for each entry (partial renaming and incorrect disk quota updates are possible in error cases). Default `0`.

The current development version will be tagged as v2.5.0 in a few weeks, any testing is greatly appreciated. Thanks

**Reply by mezuky04** [reply]
Awesome. I don't have experience with Go but I will try to build it from the source and test the parameter. Thanks for your quick answer 🙂

**Reply by drakkan** [reply]
> Awesome. I don't have experience with Go but I will try to build it from the source and test the parameter. Thanks for your quick answer slightly_smiling_face

No need to build from source you can download build artifacts for each commit from the [actions](https://github.com/drakkan/sftpgo/actions) page.
For example you can download deb/rpm, windows installer etc for the latest commit from [here](https://github.com/drakkan/sftpgo/actions/runs/4503365517). If you are on Docker just use an edge variant
